components/raftstore/src/store/fsm/peer.rs

```diff
@@ -987,36 +988,98 @@ where
+        if let Some(state) = self
+            .ctx
+            .engines
+            .kv
+            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)
+            .unwrap()
+        {
+            info!(
+                "Prepare Flashback, at now flashback persist state is: {} for {:?} ",
+                state.get_region().get_is_in_flashback(),
+                self.region(),
+            );
+            if !state.get_region().get_is_in_flashback() {
+                // callback itself to wait for the new change to be applied completely.
+                let raft_router_clone = self.ctx.router.clone();
+                let ch_clone = ch.clone();
+                let cb = Callback::write(Box::new(move |resp| {
+                    if resp.response.get_header().has_error() {
+                        ch.send(false).unwrap();
+                        error!("send flashback prepare msg failed"; "region_id" => region_id);
+                        return;
+                    }
+                    raft_router_clone
+                        .force_send(
+                            region_id,
+                            PeerMsg::SignificantMsg(SignificantMsg::PrepareFlashback(ch_clone)),
+                        )
+                        .unwrap();
+                }));
+                // set persist state by a Raft Admin Command
+                let req = {
+                    let mut request = new_admin_request(region_id, self.fsm.peer.peer.clone());
+                    let mut admin = AdminRequest::default();
+                    admin.set_cmd_type(AdminCmdType::PrepareFlashback);
+                    request.set_admin_request(admin);
+                    request
+                };
+                self.propose_raft_command(req, cb, DiskFullOpt::AllowedOnAlmostFull);
+                return;
+            }
+            // check memory state
+            if self.fsm.peer.flashback_state.is_none() {
+                self.fsm.peer.flashback_state = Some(FlashbackState::new(Some(ch)));
+            }
         }
-        self.fsm.peer.flashback_state = Some(FlashbackState::new(ch));
         // Let the leader lease to None to ensure that local reads are not executed.
         self.fsm.peer.leader_lease_mut().expire_remote_lease();
         self.fsm.peer.maybe_finish_flashback_wait_apply();
     }
 
-    fn on_finish_flashback(&mut self) {
+    fn on_finish_flashback(&mut self, ch: Sender<bool>) {
+        let region_id = self.region().get_id();
         info!(
             "finish flashback";
-            "region_id" => self.region().get_id(),
+            "region_id" => region_id,
             "peer_id" => self.fsm.peer.peer_id(),
         );
+        let ch_clone = ch.clone();
+        let cb = Callback::write(Box::new(move |resp| {
+            if resp.response.get_header().has_error() {
+                ch_clone.send(false).unwrap();
+                error!("send flashback prepare msg failed"; "region_id" => region_id);
```

Comment:
Should return here. Otherwise, the channel will be notified twice.

Reply:
U r right...I will add it.